core/src/main/java/org/apache/accumulo/core/summary/Gatherer.java

```diff
@@ -0,0 +1,513 @@
+    final Set<String> failedFiles;
+
+    public ProcessedFiles() {
+      this.summaries = new SummaryCollection();
+      this.failedFiles = new HashSet<>();
+    }
+  }
+
+  private class FilesProcessor implements Callable<ProcessedFiles> {
+
+    HostAndPort location;
+    Map<String,List<TRowRange>> allFiles;
+    private TInfo tinfo;
+
+    public FilesProcessor(TInfo tinfo, HostAndPort location, Map<String,List<TRowRange>> allFiles) {
+      this.location = location;
+      this.allFiles = allFiles;
+      this.tinfo = tinfo;
+    }
+
+    @Override
+    public ProcessedFiles call() throws Exception {
+      ProcessedFiles pfiles = new ProcessedFiles();
+
+      Client client = null;
+      try {
+        client = ThriftUtil.getTServerClient(location, ctx);
+        // partition files into smaller chunks so that not too many are sent to a tserver at once
+        for (Map<String,List<TRowRange>> files : partition(allFiles, 500)) {
+          if (pfiles.failedFiles.size() > 0) {
+            // there was a previous failure on this tserver, so just fail the rest of the files
+            pfiles.failedFiles.addAll(files.keySet());
+            continue;
+          }
+
+          TSummaries tSums = null;
+          try {
+            tSums = client.getSummariesFromFiles(tinfo, ctx.rpcCreds(), getRequest(), files);
+          } catch (TApplicationException tae) {
+            throw tae;
+          } catch (TTransportException e) {
+            pfiles.failedFiles.addAll(files.keySet());
+            continue;
+          }
+
+          pfiles.summaries.merge(new SummaryCollection(tSums), factory);
+        }
+      } finally {
+        ThriftUtil.returnClient(client);
+      }
+
+      return pfiles;
+    }
+  }
+
+  /**
+   * This methods reads a subset of file paths into memory and groups them by location. Then it request sumaries for files from each location/tablet server.
+   */
+  public SummaryCollection processPartition(int modulus, int remainder) throws TableNotFoundException, AccumuloException, AccumuloSecurityException {
+
+    Predicate<String> hashingFileSelector = file -> Math.abs(Hashing.murmur3_32().hashString(file).asInt()) % modulus == remainder;
+
+    Map<String,Map<String,List<TRowRange>>> filesGBL = getFilesGroupedByLocation(hashingFileSelector);
+
+    SummaryCollection partitionSummaries = new SummaryCollection();
+
+    ExecutorService execSrv = Executors.newCachedThreadPool();
+    try {
+      TInfo tinfo = Tracer.traceInfo();
+      while (true) {
+        Set<String> failedFiles = new HashSet<>();
+
+        List<Future<ProcessedFiles>> futures = new ArrayList<>();
+        for (Entry<String,Map<String,List<TRowRange>>> entry : filesGBL.entrySet()) {
+          HostAndPort location = HostAndPort.fromString(entry.getKey());
+          Map<String,List<TRowRange>> allFiles = entry.getValue();
+
+          Future<ProcessedFiles> future = execSrv.submit(new FilesProcessor(tinfo, location, allFiles));
```

Comment:
From `Executors.newCachedThreadPool()` javadoc:

> These pools will typically improve the performance of programs that execute many short-lived asynchronous tasks

But, in `FilesProcessor` aren't we running an RPC without timeout? Seem to be competing goals (and potentially keep us to hold on to a lot of threads if RPCs are delayed).